Fix format of values in confusion matrix plot.

Rick-Mackenbach opened the pull request:
#### Reference Issues/PRs
In reference to this, https://github.com/scikit-learn/scikit-learn/issues/16127 


#### What does this implement/fix? Explain your changes.
The formatting of values in the plot of the confusion matrix were set to '.2g' if no other format was specified. As of now, if no format is provided by the user, the format will be 'd' if the value is <1e7, and '.2g' otherwise.

Commit:
"Changes to the values_format for plotting the confusion_matrix. This
fix will now show values up to 1e7 in integer values, and afterwards
in .2g"

Commit:
"Changes to the values_format for plotting the confusion_matrix. This
fix will now show values up to 1e7 in integer values, and afterwards
in .2g"

Commit:
"Fixes for confusion_matrix values_format."

Commit:
"Fixes for confusion_matrix values_format."

Commit:
"<= value for 0"

Commit:
"<= value for 0"

Commit:
"Changed back to == 0"

Commit:
"Changed back to == 0"

Commit:
"Import math log10 for linting error"

Commit:
"Import math log10 for linting error"

Commit:
"Linting and hard-coded value checks now working"

Commit:
"Linting and hard-coded value checks now working"

Commit:
"Use log(10) instead of 1e7 (python version error?)"

Commit:
"Use log(10) instead of 1e7 (python version error?)"

Commit:
"If cmd.type == f test included"

Commit:
"If cmd.type == f test included"

Commit:
"Added type test in the correct place"

Commit:
"Added type test in the correct place"

Commit:
"Different IF statement, see if it fixes azure pipeline errors"

Commit:
"Different IF statement, see if it fixes azure pipeline errors"

Commit:
"Changed order and use np.log10 ofcourse"

Commit:
"Changed order and use np.log10 ofcourse"

Commit:
"Removed brackets"

Commit:
"Removed brackets"

Commit:
"Added testing in test_plot_confusion_matrix (values_format)"

Commit:
"Added testing in test_plot_confusion_matrix (values_format)"

Commit:
"Wrong value format"

Commit:
"Wrong value format"

Commit:
"Changed the test_plot_confusion_matrix."

Commit:
"Changed the test_plot_confusion_matrix."

Commit:
"Add test plot confusion matrix"

Commit:
"Add test plot confusion matrix"

Commit:
"Added extra test cases for test_plot_confusion_matrix"

Commit:
"Added extra test cases for test_plot_confusion_matrix"

Commit:
"Fixed testing values"

Commit:
"Fixed testing values"

Commit:
"Added argument to the test function, duh"

Commit:
"Added argument to the test function, duh"

Rick-Mackenbach commented:
@jnothman Hey Joel, I made the fix name a bit clearer now. 

Could you help me understand why the codecov/patch and codecov/project tests are not passing?

Commit:
"New tests to also test if for values > 1e7"

Commit:
"New tests to also test if for values > 1e7"

Commit:
"New test version, made sure the values are now loaded into the model
to check for array_equal."

Commit:
"New test version, made sure the values are now loaded into the model
to check for array_equal."

Commit:
"Improved test file"

Commit:
"Improved test file"

Commit:
"New changes"

Commit:
"New changes"

Commit:
"Fixed the ConfusionMatrixDisplay class, where values where not
counted in the if statement."

Commit:
"Fixed the ConfusionMatrixDisplay class, where values where not
counted in the if statement."

Rick-Mackenbach commented:
Well that certainly was a journey, but it seems to be working now.

NicolasHug commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,38 @@ def plot(self, include_values=True, cmap='viridis',
-                                           ha="center", va="center",
-                                           color=color)
+
+            if (values_format is None):
```
No parenthesis

NicolasHug commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,38 @@ def plot(self, include_values=True, cmap='viridis',
+                    self.text_[i, j] = ax.text(
+                        j, i, format(cm[i, j], values_format),
+                        ha="center", va="center",
+                        color=color)
```
These lines are duplicated.

Let's put the `if value_format is None` condition inside of the for loop to avoid duplication.

NicolasHug commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,38 @@ def plot(self, include_values=True, cmap='viridis',
         self.im_ = ax.imshow(cm, interpolation='nearest', cmap=cmap)
         self.text_ = None
+        if (values_format is None and cm.dtype.kind == "f"):
+            values_format = '.2g'
```
merge this condition with the rest below? It's better to have all related logic at the same place

NicolasHug commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -247,20 +246,47 @@ def test_confusion_matrix_pipeline(pyplot, clf, data, n_classes):
-    text_text = np.array([
-        t.get_text() for t in disp.text_.ravel()])
-    assert_array_equal(expected_text, text_text)
+    if values_format is None:
```
Having a big if/else statement like this kinda defeats the purpose of parametrizing the test.

Maybe keep this test unchanged and create a new one `test_confusion_matrix_text_format_default` which would only check the default behavior.

NicolasHug reviewed (commented):
Thanks @Rick-Mackenbach , made a few comments

Just FYI, you don't need to close and open new PR, you can simply push changes to your already-existing PR. It makes it easier for us to keep track of it ;)

Commit:
"Fixing of loop and test"

Commit:
"Fixing of loop and test"

Commit:
"Changed if statement in the for loop and new test function"

Commit:
"Changed if statement in the for loop and new test function"

Commit:
"Changes to test function and added the if condition in the for-loop
inside the ConfusionMatrixDisplay."

Commit:
"Changes to test function and added the if condition in the for-loop
inside the ConfusionMatrixDisplay."

Commit:
"Removed the assert equal size statement."

Commit:
"Removed the assert equal size statement."

Rick-Mackenbach commented:
@NicolasHug ✔

NicolasHug commented:
It seems like there are a few unaddressed comments @Rick-Mackenbach https://github.com/scikit-learn/scikit-learn/pull/16159#discussion_r368976481 and https://github.com/scikit-learn/scikit-learn/pull/16159#discussion_r368976859

Commit:
"Forgot to add the improved ConfusionMatrixDisplay"

Commit:
"Forgot to add the improved ConfusionMatrixDisplay"

Commit:
"Cleaned test file and now all inside the for-loop."

Commit:
"Cleaned test file and now all inside the for-loop."

NicolasHug commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,37 @@ def plot(self, include_values=True, cmap='viridis',
+                        j, i, format(cm[i, j], values_format),
+                        ha="center", va="center",
+                        color=color)
+                    values_format = None
```
Instead of doing this trick, just use `fmt` instead of `values_format` and remove the else condition below which is redundant

Rick-Mackenbach replied:
Not sure what you mean by this, declare a new variable 'fmt' and use that in the values_format place?

NicolasHug replied:
something like

```py
fmt = values_format
if fmt is None:
	fmt = ...


# and then
... format(cm[i, j], fmt), ...
```

Rick-Mackenbach replied:
Yea, that makes more sense, thanks.

Commit:
"Declare variable before the if-statement check"

Commit:
"Declare variable before the if-statement check"

Commit:
"Made the if-statement cleaner."

Commit:
"Made the if-statement cleaner."

jnothman commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,28 @@ def plot(self, include_values=True, cmap='viridis',
+                fmt = values_format
+
+                if values_format is None:
+                    if cm.dtype.kind == "f" or cm[i, j] >= 1e7:
```
If the matrix is normalised, won't its values be <1? Should we be showing normalized values as %?

Rick-Mackenbach replied:
Yes that's true <= 1 (in the case all predictions are correct). Wouldn't it be easier to let this be specified by the user in the case of normalization? The edge cases to choose if % or 'd'?

Also, the way plot_confusion_matrix is defined now is by letting users pass in an estimator, X and y_true. 

Wouldn't it make more sense to use y_pred and y_true and pass those as arguments?

NicolasHug replied:
at this point it might be easier to literally check the length of the formatted string with both options and just use the shortest one

Rick-Mackenbach replied:
@NicolasHug The format will then go to '.2g' at values => 100 (which is the same as before)

Rick-Mackenbach replied:
@NicolasHug Values will then change to '.2g' format once >= 1e6

thomasjpfan replied:
Currently this does not always pick the shorter one. if `cm[i, j] == int(1e6)`, `fmt='d'` will be used, which results in: `1000000`, while `fmt='.2g'` gives `'1e+06'`

NicolasHug commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,31 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+            return '.2g'
+
+    X, y = data
+    cm2 = confusion_matrix([1, 1]*5000000, [0, 0]*5000000)
```
no need to create such a huge list

NicolasHug commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,31 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+
+
+@pytest.mark.parametrize("values_format", [None])
+def test_confusion_matrix_standard_format(pyplot, data, y_pred, n_classes,
```
I think the test would be more helpful if it were checking explicit formatted values.

i.e. pass in 1123124, make sure it's rendered as 1123124. Pass in .345, make sure it's rendered as .35...

It will make it easier for us to verify that it works as expected

jnothman replied:
I am confused by what this latest change was trying to do. I liked how you had `assert_array_equal(expected_text, text_text)`, but I didn't like how expected_text was constructed by formatting the numbers, rather than as a list of strings manually entered into the test.

jnothman replied:
I am confused by what this latest change was trying to do. I liked how you had `assert_array_equal(expected_text, text_text)`, but I didn't like how expected_text was constructed by formatting the numbers, rather than as a list of strings manually entered into the test.

Rick-Mackenbach replied:
Ahh ok, I see. Sorry I misunderstood it and thought you meant just directly check these values formatting.

Commit:
"Updated test file, and added the statement where if the length of the
'.2g' value is larger than the length of the '.d' value, it will pick
the shortest one."

Commit:
"Updated test file, and added the statement where if the length of the
'.2g' value is larger than the length of the '.d' value, it will pick
the shortest one."

Commit:
"Changed back to values larger than 1e7"

Commit:
"Changed back to values larger than 1e7"

Commit:
"Test if the failed tests on azure pipelines are because of my edits?"

Commit:
"Test if the failed tests on azure pipelines are because of my edits?"

jnothman commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,51 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+    # Plot the values to see if it works
+    disp.plot()
+
+    expected_text = np.array([format(v, test_function_format(v))
```
I think we are better off explicitly writing out the expected strings than reproducing the logic from the code. This makes our expectations more explicit and so easier for the reader to confirm that the test is working towards desirable behaviour

jnothman replied:
Are you able to address this, @Rick-Mackenbach? I am very uncomfortable seeing `test_function_format` in the test.

jnothman replied:
Sorry, it seems you did not understand my comment. I mean that `test_function_format` reproduces logic from the implementation. Rather we should be checking the output against expected strings directly.

Commit:
"Improved test file, and included expected values from the test."

Commit:
"Improved test file, and included expected values from the test."

Rick-Mackenbach commented:
I think this one should about do the trick, what do you think?

Commit:
"Improved test of formats (explicitly check strings)"

Commit:
"Improved test of formats (explicitly check strings)"

Commit:
"Modified the test function for test_plot_confusion_matrix"

Commit:
"Modified the test function for test_plot_confusion_matrix"

Commit:
"Improved test file, now check for literal string array equal."

Commit:
"Improved test file, now check for literal string array equal."

Commit:
"Use ravel instead of flatten"

Commit:
"Use ravel instead of flatten"

jnothman commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,37 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+    test = []
+    for i in plotted_text_values.ravel():
+        test.append(i.get_text())
+    test = np.asarray(test)
```
Avoid making these into arrays. Just use a bare `assert` statement to compare the two lists, please

Rick-Mackenbach replied:
✔

Commit:
"Removed array and used list only with bare assert statement."

Commit:
"Removed array and used list only with bare assert statement."

jnothman commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,35 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+    test = []
+    for i in plotted_text_values.ravel():
+        test.append(i.get_text())
+    assert test == ['1e+07', '0', '29', '123123']
```
if these expected values are hardcoded in the test, it no loner makes sense for values_format to be a parameter. Alternatively, if you want to make values_format a parameter, the expected values must also be parameters.

Rick-Mackenbach replied:
Sorry, you're totally right about that, seems silly to parametrize this indeed.

jnothman commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,35 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+    disp = plot_confusion_matrix(fitted_clf, X, y,
+                                 include_values=True,
+                                 values_format=values_format)
+    disp.confusion_matrix = cm
```
This approach to modifying the attribute is a pretty strange usage. We don't do it elsewhere in tests or examples. Why not just use `ConfusionMatrixDisplay(cm, [False, True]).plot(values_format=values_format)`?

Rick-Mackenbach replied:
Yea the code becomes much simpler using this line, sorry about that. I'm still new to contributing to projects, so sorry it takes some extra time :).

I just took this testing as it was how I did it in jupy notebooks.

Commit:
"Cleaned up the test file, no parameterize and cleaner code using
ConfusionMatrixDisplay(cm, ...) "

Commit:
"Cleaned up the test file, no parameterize and cleaner code using
ConfusionMatrixDisplay(cm, ...) "

Commit:
"Added pyplot to the function parameter so it passes tests."

Commit:
"Added pyplot to the function parameter so it passes tests."

Commit:
"Use values_format in parameter test, to see if it resolves
unit test error on github side."

Commit:
"Use values_format in parameter test, to see if it resolves
unit test error on github side."

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,26 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+    assert test == ['1e+07', '0', '29', '123123']
+
+    cm = np.array([[0.1, 10], [100, 0.5]])
+    plotted_text_values = ConfusionMatrixDisplay(cm, [False, True]).plot(
```
nit:

```suggestion
    plotted_text = ConfusionMatrixDisplay(cm, [False, True]).plot(
```

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,26 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
+                                            values_format=values_format).text_
+    # Values should now formatted as '.2g',
+    # Values are have two dec places max, and =>100 becomes e+02.
+    test = []
```
Nit 
```py
    text_values = [t.get_text() for t in plotted_text.ravel()]
```

thomasjpfan commented on sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py:
```diff
@@ -264,3 +263,26 @@ def test_confusion_matrix_text_format(pyplot, data, y_pred, n_classes,
     assert_array_equal(expected_text, text_text)
+
+
+@pytest.mark.parametrize("values_format", [None])
```
This parametrize is not needed anymore.

Commit:
"Shorten the test file, rename of some variables, list comprehension"

Commit:
"Shorten the test file, rename of some variables, list comprehension"

Commit:
"Changed to 1e6"

Commit:
"Changed to 1e6"

Commit:
"Now picks the shortest length format."

Commit:
"Now picks the shortest length format."

Rick-Mackenbach commented:
Any insight into why this ( ci/circleci: doc-min-dependencies — Your tests failed on CircleCI ) test is failing?

thomasjpfan commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,30 @@ def plot(self, include_values=True, cmap='viridis',
-                                           color=color)
+
+                fmt = values_format
+                if values_format is None:
```
Maybe this:

```py
if values_format is None:
    text_cm = format(cm[i, j], '.2g')
    if cm.dtype.kind != 'f':
        text_d = format(cm[i, j], 'd')
        if len(text_d) < len(text_cm):
            text_cm = text_d
else:
    text_cm = format(cm[i, j], values_format)
```

The test would need to updated to reflect this behavior.

thomasjpfan replied:
I think trying `.2g` and then trying `d` if applicable, is slightly clearer

This suggestion is to address https://github.com/scikit-learn/scikit-learn/pull/16159#discussion_r370265170

```py
if values_format is None:
    text_cm = format(cm[i, j], '.2g')
    if cm.dtype.kind != 'f':
        text_d = format(cm[i, j], 'd')
        if len(text_d) < len(text_cm):
            text_cm = text_d
else:
    text_cm = format(cm[i, j], values_format)
```

The test would need to updated to reflect this behavior.

thomasjpfan reviewed (commented):
Merging with master may help with the doc-min error

Commit:
"Small changes in confusion plot values"

Commit:
"Small changes in confusion plot values"

Commit:
Merge branch 'confusion_fix'

Commit:
Merge branch 'confusion_fix' of https://github.com/Rick-Mackenbach/scikit-learn into confusion_fix

jnothman commented:
I don't think that merge worked well... The diff now includes all changes.

 You should have been able to do something like:
```
git checkout master
git pull https://github.com/scikit-learn/scikit-learn master
git checkout my_branch
git merge master
git push
```

not sure that'll work now. You might want to reset back to your last good commit using `git reset --hard`. But it seems your commits and the commits in master are now interleaved...

It might be better to lose your history by doing something like:
```
git checkout confusion_fix
git reset --hard master
git checkout master sklearn/metrics/_plot/confusion_matrix.py sklearn/metrics/_plot/tests/test_plot_confusion_matrix.py  # are these the only files you intended to modify in this PR??
git status
git commit
```

Commit:
"Fix of wrong merge"

Commit:
Merge branch 'confusion_fix' of https://github.com/Rick-Mackenbach/scikit-learn into confusion_fix

Rick-Mackenbach commented:
Thanks @jnothman for the fix!

jnothman commented:
I hope it kept all your changes!

Rick-Mackenbach commented:
> I hope it kept all your changes!

Seems like it did from what I can see :)

Commit:
"Small change in test file"

thomasjpfan reviewed (commented):
This looks almost done!

Commit:
"Updated the logic for interpretability. Also small changes to test"

Rick-Mackenbach commented:
@thomasjpfan Not sure if this is what you meant?

thomasjpfan commented on sklearn/metrics/_plot/confusion_matrix.py:
```diff
@@ -83,22 +83,30 @@ def plot(self, include_values=True, cmap='viridis',
+                                                        format(cm[i, j], 'd')):
+                            fmt = '.2g'
+
+                self.text_[i, j] = ax.text(
```
Sorry I was not more clear. The current PR may call `format` three times. I am trying to push for calling it at most twice:

```py
if values_format is None:
    text_cm = format(cm[i, j], '.2g')
    if cm.dtype.kind != 'f':
        text_d = format(cm[i, j], 'd')
        if len(text_d) < len(text_cm):
            text_cm = text_d
else:
    text_cm = format(cm[i, j], values_format)

self.text_[i, j] = ax.text(j, i, text_cm, ...)
```

thomasjpfan reviewed (commented):
Thank you for your continued work and patience on this PR.

Commit:
"Do the formatting of values in the if-statement, so shorter code :)"

Rick-Mackenbach commented:
@thomasjpfan My pleasure, thanks for you continuous feedback on this :). 

I wasn't looking correctly regarding the correction, I see now that the formatting is done in the if-statement, and you just pass this formatted text to self.text_

Should be correct now 👍

Commit:
"Changed test file to test for 'd' values"

thomasjpfan reviewed (approved):
LGTM

thomasjpfan commented:
Please add an entry to the change log at `doc/whats_new/v0.23.rst` with a `API` tag. Like the other entries there, please reference this pull request with `:pr:` and credit yourself (and other contributors if applicable) with `:user:`.

Commit:
"Added changes to the v0.23 what's news"

Commit:
"Modified doc/whats_new/v0.23.rst"

Rick-Mackenbach commented:
@thomasjpfan Not sure why the tests have failed here, could you help me out a bit? :)

thomasjpfan committed:
Merge remote-tracking branch 'upstream/master' into pr/16159

thomasjpfan commented:
I pushed a change to fix up the PR. It should be ready now.

jnothman commented:
Thanks @Rick-Mackenbach!
